posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,256 @@
+            for team_id, old_person_ids in inputs.to_override.items():
+                # We could use executemany instead of looping, but in its current
+                # implementation executemany is not faster than execute in a loop.
+                # https://www.psycopg.org/docs/cursor.html#cursor.executemany
+                cursor.execute(
+                    DELETE_FROM_PERSON_OVERRIDE_TABLES,
+                    {
+                        "team_id": team_id,
+                        "overriden_persons": [UUID(old_person_id) for old_person_id in old_person_ids],
+                    },
+                )
+
+
+@dataclass
+class SquashPersonOverridesInputs:
+    """Inputs for the SquashPersonOverrides workflow."""
+
+    clickhouse_database: str = "default"
+    postgres_database: str = "posthog"
+    join_table_name: str = "person_overrides_join"
+    partition_ids: list[str] | None = None
+    last_n_months: int | None = None
+
+    def iter_partition_ids(self) -> Iterator[str]:
+        """Iterate over configured partition ids.
+
+        If partition_ids is set, then we will just yield from that.
+        Otherwise, we compute the partition keys for the last_n_months.
+        """
+        if not self.partition_ids and not self.last_n_months:
+            # Checking for falsy values also ensures we are not left with [] and 0
+            raise ValueError("Either partition_ids or last_n_months must be set")
+
+        if self.partition_ids:
+            yield from self.partition_ids
+            return
+
+        current_month = datetime.now()
+
+        for _ in range(self.last_n_months):
+            current_month = current_month.replace(day=1)
+
+            yield current_month.strftime("%Y%m")
+
+            current_month = current_month - timedelta(days=1)
+
+
+@workflow.defn
+class SquashPersonOverrides:
+    """Workflow to squash outstanding person overrides into events.
+
+    Squashing refers to the process of updating the person_id associated with an event
+    to match the new id assigned via a person override. This process must be done
+    regularly to control the size of the person_overrides table.
+    """
+
+    @workflow.run
+    async def run(self, inputs: SquashPersonOverridesInputs):
+        """Workflow implementation to squash person overrides into events table."""
+        for partition_id in inputs.iter_partition_ids():
+            workflow.logger.info("Running squash workflow for partition %s", partition_id)
+            to_override = await workflow.execute_activity(
+                select_persons_to_squash,
+                QueryInputs(partition_id, database=inputs.clickhouse_database),
+                start_to_close_timeout=timedelta(minutes=5),
+            )
+
+            if not to_override:
+                workflow.logger.info("No persons to squash found in partition %s", partition_id)
+                continue
+
+            workflow.logger.info("Squashing %s persons", len(to_override))
```

Comment:
we might be squashing more recently added entries too (that weren't in `to_override`), no? This should be fine as person overrides are are fine to squash partially, but could be confusing.

Reply:
Yeah, `to_override` represents the person overrides that are safe to delete as we found them before starting the squash. During the squash we may do more work, and the overrides in this extra work will not be cleaned up during this run.

This should be safe as once the overrides are squashed we will not be querying the table during joins, and whatever wasn't cleaned up this time would be cleaned up the next time we run the squash job. The trade off for this is simpler queries.

I can clarify a bit more in the logs that this is an approximation and attempt to get the final number from the actual squash query.

Reply:
Agh, looks like ClickHouse doesn't return number of updated or inserted rows, like Postgres: https://github.com/ClickHouse/ClickHouse/issues/19274.

I was going to use that for a more exact logging of the amount of work we did. We could query the join table for a size check, but maybe we do that later if we deem it necessary. I will update the log lines so they clarify the number refers to number of persons safe to delete not actual work.

Reply:
yup just clarifying the log lines or adding comments would be fine.